File: android/org/ngui/Android.java
/* ***** BEGIN LICENSE BLOCK *****
 * Distributed under the BSD license:
 *
 * Copyright (c) 2015, xuewen.chu
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of xuewen.chu nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL xuewen.chu BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * ***** END LICENSE BLOCK ***** */

package org.ngui;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.BatteryManager;
import android.os.Build;
import android.os.Debug;
import android.telephony.TelephonyManager;
import java.util.List;
import java.util.Locale;
import android.app.ActivityManager;
import android.util.Log;

public class Android {

	private static String TAG = "Ngui";
	private static NGUIActivity activity = null;
	private static int battery_status = BatteryManager.BATTERY_STATUS_UNKNOWN;
	private static boolean is_ac_power_connected = false;
	private static NGUIActivity.PrivateAPI api = null;

	private static BroadcastReceiver receiver = new BroadcastReceiver() {
		@Override
		public void onReceive(Context context, Intent intent) {
			if ( intent.getAction() == Intent.ACTION_BATTERY_CHANGED ) {
				Log.d(TAG, "ACTION_BATTERY_CHANGED");
				battery_status = intent.getIntExtra("status", BatteryManager.BATTERY_STATUS_UNKNOWN);
			} else if ( intent.getAction() == Intent.ACTION_POWER_CONNECTED ) {
				is_ac_power_connected = true;
				Log.d(TAG, "ACTION_POWER_CONNECTED");
			} else {
				Log.d(TAG, "ACTION_POWER_DISCONNECTED");
				is_ac_power_connected = false;
			}
		}
	};

	public static void initialize(NGUIActivity act, NGUIActivity.PrivateAPI api) {
		if ( !act.equals(activity) ) {
			activity = act;
			Android.api = api;
			activity.registerReceiver(receiver, new IntentFilter(Intent.ACTION_BATTERY_CHANGED));
			activity.registerReceiver(receiver, new IntentFilter(Intent.ACTION_POWER_CONNECTED));
			activity.registerReceiver(receiver, new IntentFilter(Intent.ACTION_POWER_DISCONNECTED));
		}
	}

	public static void uninitialize(NGUIActivity act) {
		if ( act.equals(activity) ) {
			activity.unregisterReceiver(receiver);
			activity = null;
			api = null;
		}
	}

	// -------------------- gui --------------------

	private static void ime_keyboard_open(final boolean clear, final int type, final int return_type) {
		activity.post(new Runnable() {
			public void run() { api.ime_keyboard_open(clear, type, return_type); }
		});
	}

	private static void ime_keyboard_can_backspace(final boolean can_backspace, final boolean can_delete) {
		activity.post(new Runnable() {
			public void run() { api.ime_keyboard_can_backspace(can_backspace, can_delete); }
		});
	}

	private static void ime_keyboard_close() {
		activity.post(new Runnable() {
			public void run() { api.ime_keyboard_close(); }
		});
	}

	private static void keep_screen(final boolean value) {
		activity.post(new Runnable() {
			public void run() {
				api.keep_screen(value);
			}
		});
	}

	private static int  get_status_bar_height() {
		return api.get_status_bar_height();
	}

	private static void set_visible_status_bar(final boolean visible) {
		activity.post(new Runnable() {
			public void run() {
				api.set_visible_status_bar(visible);
			}
		});
	}

	private static void set_status_bar_style(final int style) {
		activity.post(new Runnable() {
			public void run() {
				api.set_status_bar_style(style);
			}
		});
	}

	private static void request_fullscreen(final boolean fullscreen) {
		activity.post(new Runnable() {
			public void run() {
				api.request_fullscreen(fullscreen);
			}
		});
	}

	private static int  get_orientation() {
		return api.get_orientation();
	}

	private static void set_orientation(final int orientation) {
		activity.post(new Runnable() {
			public void run() {
				api.set_orientation(orientation);
			}
		});
	}

	private static float get_display_scale() {
		return api.get_display_scale();
	}

	private static boolean is_screen_on() {
		return api.is_screen_on();
	}

	private static void set_volume_up() {
		activity.post(new Runnable() {
			public void run() {
				api.set_volume_up();
			}
		});
	}

	private static void set_volume_down() {
		activity.post(new Runnable() {
			public void run() {
				api.set_volume_down();
			}
		});
	}

	private static void open_url(final String url) {
		activity.post(new Runnable() {
			public void run() {
				api.open_url(url);
			}
		});
	}

	private static void send_email(final String recipient,
																 final String subject,
																 final String cc, final String bcc, final String body) {
		activity.post(new Runnable() {
			public void run() {
				api.send_email(recipient, subject, cc, bcc, body);
			}
		});
	}

	// -------------------- util --------------------

	private static String start_path() {
		return activity.startPath();
	}
	
	private static String package_code_path() {
		return activity.getPackageCodePath();
	}

	private static String files_dir_path() {
		return activity.getFilesDir().getPath();
	}

	private static String cache_dir_path() {
		return activity.getCacheDir().getPath();
	}

	private static String version() {
		return Build.VERSION.RELEASE;
	}

	private static String brand() {
		return Build.BRAND;
	}

	private static String subsystem() {
		return Build.MODEL;
	}

	private static int network_status() {
		ConnectivityManager m = (ConnectivityManager)
						activity.getSystemService(Context.CONNECTIVITY_SERVICE);
		NetworkInfo info = m.getActiveNetworkInfo();

		if ( info != null && info.isAvailable() ) {
			int type = info.getType();
			if (type == ConnectivityManager.TYPE_WIFI) {
				return 2;
			} else if (type == ConnectivityManager.TYPE_MOBILE) {

				int sub_type = info.getSubtype();
				TelephonyManager tm = (TelephonyManager)
								activity.getSystemService(Context.TELEPHONY_SERVICE);

				if (sub_type == TelephonyManager.NETWORK_TYPE_LTE) {
					return 5; // 4G
				} else if (sub_type == TelephonyManager.NETWORK_TYPE_UMTS ||
								sub_type == TelephonyManager.NETWORK_TYPE_HSDPA ||
								sub_type == TelephonyManager.NETWORK_TYPE_EVDO_0) {
					return 5; // 3G
				} else if (sub_type == TelephonyManager.NETWORK_TYPE_GPRS ||
								sub_type == TelephonyManager.NETWORK_TYPE_EDGE ||
								sub_type == TelephonyManager.NETWORK_TYPE_CDMA) {
					return 4; // 2G
				} else {
					return 4; // 2G
				}
			}
		}
		return 0;
	}

	private static boolean is_ac_power() {
		if ( is_battery() ) {
			switch ( battery_status ) {
				case BatteryManager.BATTERY_STATUS_CHARGING: return true;
				case BatteryManager.BATTERY_STATUS_DISCHARGING: return false;
				case BatteryManager.BATTERY_STATUS_NOT_CHARGING:
				case BatteryManager.BATTERY_STATUS_FULL:
				case BatteryManager.BATTERY_STATUS_UNKNOWN: break;
			}
			return is_ac_power_connected;
		} else { // no battery
			return true;
		}
	}

	private static boolean is_battery() {
		BatteryManager bm =(BatteryManager)activity.getSystemService(Context.BATTERY_SERVICE);
		int capacity = bm.getIntProperty(BatteryManager.BATTERY_PROPERTY_CHARGE_COUNTER);
		return capacity > 0;
	}

	private static float battery_level() {
		if ( is_battery() ) {
			BatteryManager bm = (BatteryManager) activity.getSystemService(Context.BATTERY_SERVICE);
			float level = bm.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY) / 100F;
			return level;
		} else {
			return 0;
		}
	}

	private static String language() {
		return Locale.getDefault().toString();
	}

	private static ActivityManager.MemoryInfo memory_info() {
		ActivityManager am = (ActivityManager)activity.getSystemService(Context.ACTIVITY_SERVICE);
		ActivityManager.MemoryInfo info = new ActivityManager.MemoryInfo();
		am.getMemoryInfo(info);
		return info;
	}

	private static long available_memory() {
		return memory_info().availMem;
	}

	private static long memory() {
		return memory_info().totalMem;
	}

	private static long used_memory() {
		ActivityManager am = (ActivityManager)activity.getSystemService(Context.ACTIVITY_SERVICE);
		List<ActivityManager.RunningAppProcessInfo> appProcessList = am.getRunningAppProcesses();
		int pid = android.os.Process.myPid();
		for (ActivityManager.RunningAppProcessInfo appProcessInfo : appProcessList) {
			if ( appProcessInfo.pid == pid ) {
				Debug.MemoryInfo[] memoryInfo = am.getProcessMemoryInfo(new int[] { pid });
				return memoryInfo[0].dalvikPrivateDirty * 1024;
			}
		}
		return 0;
	}

}

File: src/android/android.cc
/* ***** BEGIN LICENSE BLOCK *****
 * Distributed under the BSD license:
 *
 * Copyright (c) 2015, xuewen.chu
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of xuewen.chu nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL xuewen.chu BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 * ***** END LICENSE BLOCK ***** */

#include "android.h"
#include "ftr/util/android-jni.h"

namespace ftr {

typedef JNI::MethodInfo MethodInfo;
typedef JNI::ScopeENV   ScopeENV;

class API {
 public:
	API() {
		ScopeENV env;
		clazz_              = JNI::find_clazz("org/ftr/Android");
		//clazz_build_        = JNI::find_clazz("android.os.Build");
		// utils
		version_            = JNI::find_static_method(clazz_, "version", "()Ljava/lang/String;");
		brand_              = JNI::find_static_method(clazz_, "brand", "()Ljava/lang/String;");
		subsystem_          = JNI::find_static_method(clazz_, "subsystem", "()Ljava/lang/String;");
		package_code_path_  = JNI::find_static_method(clazz_, "package_code_path", "()Ljava/lang/String;");
		cache_dir_path_     = JNI::find_static_method(clazz_, "cache_dir_path", "()Ljava/lang/String;");
		files_dir_path_     = JNI::find_static_method(clazz_, "files_dir_path", "()Ljava/lang/String;");
		// ftr
		ime_keyboard_open_  = JNI::find_static_method(clazz_, "ime_keyboard_open", "(ZII)V");
		ime_keyboard_can_backspace_ = JNI::find_static_method(clazz_, "ime_keyboard_can_backspace", "(ZZ)V");
		ime_keyboard_close_ = JNI::find_static_method(clazz_, "ime_keyboard_close", "()V");
		keep_screen_        = JNI::find_static_method(clazz_, "keep_screen", "(Z)V");
		get_status_bar_height_  = JNI::find_static_method(clazz_, "get_status_bar_height", "()I");
		set_visible_status_bar_ = JNI::find_static_method(clazz_, "set_visible_status_bar", "(Z)V");
		set_status_bar_style_  = JNI::find_static_method(clazz_, "set_status_bar_style", "(I)V");
		request_fullscreen_ = JNI::find_static_method(clazz_, "request_fullscreen", "(Z)V");
		get_orientation_    = JNI::find_static_method(clazz_, "get_orientation", "()I");
		set_orientation_    = JNI::find_static_method(clazz_, "set_orientation", "(I)V");
		get_display_scale_  = JNI::find_static_method(clazz_, "get_display_scale", "()F");
		is_screen_on_       = JNI::find_static_method(clazz_, "is_screen_on", "()Z");
		set_volume_up_      = JNI::find_static_method(clazz_, "set_volume_up", "()V");
		set_volume_down_    = JNI::find_static_method(clazz_, "set_volume_down", "()V");
		open_url_           = JNI::find_static_method(clazz_, "open_url", "(Ljava/lang/String;)V");
		send_email_         = JNI::find_static_method(clazz_, "send_email",
																									"(Ljava/lang/String;Ljava/lang/String;"
																									"Ljava/lang/String;Ljava/lang/String;"
																									"Ljava/lang/String;)V");
		start_cmd_         = JNI::find_static_method(clazz_, "start_cmd", "()Ljava/lang/String;");
		network_status_     = JNI::find_static_method(clazz_, "network_status", "()I");
		is_ac_power_        = JNI::find_static_method(clazz_, "is_ac_power", "()Z");
		is_battery_         = JNI::find_static_method(clazz_, "is_battery", "()Z");
		battery_level_      = JNI::find_static_method(clazz_, "battery_level", "()F");
		language_           = JNI::find_static_method(clazz_, "language", "()Ljava/lang/String;");
		available_memory_   = JNI::find_static_method(clazz_, "available_memory", "()J");
		memory_             = JNI::find_static_method(clazz_, "memory", "()J");
		used_memory_        = JNI::find_static_method(clazz_, "used_memory", "()J");
		clazz_ = (jclass)env->NewGlobalRef(clazz_);
		//clazz_build_ = (jclass)env->NewGlobalRef(clazz_build_);
	}

	~API() {
		ScopeENV env;
		env->DeleteGlobalRef(clazz_);
		// env->DeleteGlobalRef(clazz_build_);
	}

	//jclass    clazz_build_;
	jclass    clazz_;
	jmethodID ime_keyboard_open_;
	jmethodID ime_keyboard_can_backspace_;
	jmethodID ime_keyboard_close_;
	jmethodID keep_screen_;
	jmethodID get_status_bar_height_;
	jmethodID set_visible_status_bar_;
	jmethodID set_status_bar_style_;
	jmethodID request_fullscreen_;
	jmethodID get_orientation_;
	jmethodID set_orientation_;
	jmethodID get_display_scale_;
	jmethodID is_screen_on_;
	jmethodID set_volume_up_;
	jmethodID set_volume_down_;
	jmethodID open_url_;
	jmethodID send_email_;
	jmethodID start_cmd_;
	jmethodID package_code_path_;
	jmethodID files_dir_path_;
	jmethodID cache_dir_path_;
	jmethodID version_;
	jmethodID brand_;
	jmethodID subsystem_;
	jmethodID network_status_;
	jmethodID is_ac_power_;
	jmethodID is_battery_;
	jmethodID battery_level_;
	jmethodID language_;
	jmethodID available_memory_;
	jmethodID memory_;
	jmethodID used_memory_;
};

#define clazz_ _core->clazz_

static API* _core = nullptr;

void Android::initialize() {
	if ( !_core ) {
		_core = new API();
	}
}
// ------------------------------- gui -------------------------------
void Android::ime_keyboard_open(bool clear, int type, int return_type) {
	ScopeENV env;
	env->CallStaticVoidMethod(clazz_, _core->ime_keyboard_open_, clear, type, return_type);
}
void Android::ime_keyboard_can_backspace(bool can_backspace, bool can_delete) {
	ScopeENV env;
	env->CallStaticVoidMethod(clazz_, _core->ime_keyboard_can_backspace_, can_backspace, can_delete);
}
void Android::ime_keyboard_close() {
	ScopeENV env;
	env->CallStaticVoidMethod(clazz_, _core->ime_keyboard_close_);
}
void Android::keep_screen(bool value) {
	ScopeENV env;
	env->CallStaticVoidMethod(clazz_, _core->keep_screen_, value);
}
int Android::get_status_bar_height() {
	ScopeENV env;
	return env->CallStaticIntMethod(clazz_, _core->get_status_bar_height_);
}
void Android::set_visible_status_bar(bool visible) {
	ScopeENV env;
	env->CallStaticVoidMethod(clazz_, _core->set_visible_status_bar_, visible);
}
void Android::set_status_bar_style(int style) {
	ScopeENV env;
	env->CallStaticVoidMethod(clazz_, _core->set_status_bar_style_, style);
}
void Android::request_fullscreen(bool fullscreen) {
	ScopeENV env;
	env->CallStaticVoidMethod(clazz_, _core->request_fullscreen_, fullscreen);
}
int Android::get_orientation() {
	ScopeENV env;
	return env->CallStaticIntMethod(clazz_, _core->get_orientation_);
}
void Android::set_orientation(int orientation) {
	ScopeENV env;
	env->CallStaticVoidMethod(clazz_, _core->set_orientation_, orientation);
}
float Android::get_display_scale() {
	ScopeENV env;
	return env->CallStaticFloatMethod(clazz_, _core->get_display_scale_);
}
bool Android::is_screen_on() {
	ScopeENV env;
	return env->CallStaticBooleanMethod(clazz_, _core->is_screen_on_);
}
void Android::set_volume_up() {
	ScopeENV env;
	env->CallStaticVoidMethod(clazz_, _core->set_volume_up_);
}
void Android::set_volume_down() {
	ScopeENV env;
	env->CallStaticVoidMethod(clazz_, _core->set_volume_down_);
}
void Android::open_url(cString& url) {
	ScopeENV env;
	env->CallStaticVoidMethod(clazz_, _core->open_url_, env->NewStringUTF(*url));
}
void Android::send_email(cString& recipient,
												 cString& subject,
												 cString& cc, cString& bcc, cString& body) {
	ScopeENV env;
	env->CallStaticVoidMethod(clazz_, _core->send_email_,
														env->NewStringUTF(*recipient),
														env->NewStringUTF(*subject),
														env->NewStringUTF(*cc),
														env->NewStringUTF(*bcc),
														env->NewStringUTF(*body)
	);
}
// ------------------------------- util -------------------------------
String Android::start_cmd() {
	ScopeENV env;
	jobject obj = env->CallStaticObjectMethod(clazz_, _core->start_cmd_);
	return JNI::jstring_to_string((jstring)obj, *env);
}
String Android::package_code_path() {
	ScopeENV env;
	jobject obj = env->CallStaticObjectMethod(clazz_, _core->package_code_path_);
	return JNI::jstring_to_string((jstring)obj, *env);
}
String Android::files_dir_path() {
	ScopeENV env;
	jobject obj = env->CallStaticObjectMethod(clazz_, _core->files_dir_path_);
	return JNI::jstring_to_string((jstring)obj, *env);
}
String Android::cache_dir_path() {
	ScopeENV env;
	jobject obj = env->CallStaticObjectMethod(clazz_, _core->cache_dir_path_);
	return JNI::jstring_to_string((jstring)obj, *env);
}
String Android::version() {
	ScopeENV env;
	jobject obj = env->CallStaticObjectMethod(clazz_, _core->version_);
	return JNI::jstring_to_string((jstring)obj, *env);
}
String Android::brand() {
	ScopeENV env;
	jobject obj = env->CallStaticObjectMethod(clazz_, _core->brand_);
	return JNI::jstring_to_string((jstring)obj, *env);
}
String Android::subsystem() {
	ScopeENV env;
	jobject obj = env->CallStaticObjectMethod(clazz_, _core->subsystem_);
	return JNI::jstring_to_string((jstring)obj, *env);
}
int Android::network_status() {
	ScopeENV env;
	return env->CallStaticIntMethod(clazz_, _core->network_status_);
}
bool Android::is_ac_power() {
	ScopeENV env;
	return env->CallStaticBooleanMethod(clazz_, _core->is_ac_power_);
}
bool Android::is_battery() {
	ScopeENV env;
	return env->CallStaticBooleanMethod(clazz_, _core->is_battery_);
}
float Android::battery_level() {
	ScopeENV env;
	return env->CallStaticFloatMethod(clazz_, _core->battery_level_);
}
String Android::language() {
	ScopeENV env;
	return JNI::jstring_to_string((jstring)env->CallStaticObjectMethod(clazz_, _core->language_), *env);
}
uint64 Android::available_memory() {
	ScopeENV env;
	return env->CallStaticLongMethod(clazz_, _core->available_memory_);
}
uint64 Android::memory() {
	ScopeENV env;
	return env->CallStaticLongMethod(clazz_, _core->memory_);
}
uint64 Android::used_memory() {
	ScopeENV env;
	return env->CallStaticLongMethod(clazz_, _core->used_memory_);
}

}